blueocean-dashboard/src/main/js/components/karaoke/components/Pipeline.jsx

```diff
@@ -0,0 +1,272 @@
+import React, { Component, PropTypes } from 'react';
+import { logging, sseConnection, calculateLogView } from '@jenkins-cd/blueocean-core-js';
+import Extensions from '@jenkins-cd/js-extensions';
+import { observer } from 'mobx-react';
+import debounce from 'lodash.debounce';
+import { QueuedState, NoSteps } from './QueuedState';
+import { KaraokeService } from '../index';
+import LogToolbar from './LogToolbar';
+import Steps from './Steps';
+import FreeStyle from './FreeStyle';
+
+import { KaraokeConfig } from '../';
+const logger = logging.logger('io.jenkins.blueocean.dashboard.karaoke.Pipeline');
+
+@observer
+export default class Pipeline extends Component {
+    constructor(props) {
+        super(props);
+        this.listener = {};
+        this.sseEventHandler = this.sseEventHandler.bind(this);
+        // query parameter before preference
+        this.classicLog = calculateLogView(props) || KaraokeConfig.getPreference('runDetails.logView').value === 'classic';
+        this.showPending = KaraokeConfig.getPreference('runDetails.pipeline.showPending').value !== 'never'; // Configure flag to show pending or not
+        this.karaoke = KaraokeConfig.getPreference('runDetails.pipeline.karaoke').value === 'never' ? false : props.augmenter.karaoke; // initial karaoke state
+        this.updateOnFinish = KaraokeConfig.getPreference('runDetails.pipeline.updateOnFinish').value;
+        this.stopOnClick = KaraokeConfig.getPreference('runDetails.pipeline.stopKaraokeOnAnyNodeClick').value === 'always';
+    }
+    componentWillMount() {
+        // starting pipeline service when we have an augmenter
+        if (this.props.augmenter) {
+            const { augmenter, params: { node } } = this.props;
+            this.pager = KaraokeService.pipelinePager(augmenter, { node });
+        }
+        // get sse listener to react on the different in sse events
+        this.listener.ssePipeline = sseConnection.subscribe('pipeline', this.sseEventHandler);
+        this.listener.sseJob = sseConnection.subscribe('job', this.sseEventHandler);
```

Comment:
I think for added safety these subscriptions should be made in `componentDidMount`, or perhaps add a "cleanup" function that's called first, and will unsubscribe from any listeners that exist.

Per Abramov, in future versions of React `componentWillMount` could be called multiple times and this might leak SSE listeners.

see: http://stackoverflow.com/questions/41612200/in-react-js-should-i-make-my-initial-network-request-in-componentwillmount-or-co